cocina_level2_prs.rb

```diff
@@ -0,0 +1,185 @@
+    result = create_branch
+    result = update_gems unless result_failure?('create_branch', result)
+    result = local_git_commit unless result_failure?('update_gems', result)
+    result = git_push_origin unless result_failure?('local_git_commit', result)
+    create_pr if result.include?(EXPECTED_PUSH_MESSAGE)
+  end
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def create_branch
+  ErrorEmittingExecutor.execute("git checkout -B #{BRANCH_NAME}")
+
+  # Ensure local branch matches any existing upstream branch
+  ErrorEmittingExecutor.execute('git reset --hard', exit_on_error: true)
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def update_gems
+  gemfile = File.read('Gemfile')
+  %w[cocina-models dor-services-client sdr-client].each do |gem_name|
+    ErrorEmittingExecutor.execute("bundle update #{gem_name}") if gemfile.include?(gem_name)
+  end
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def local_git_commit
+  ErrorEmittingExecutor.execute("git commit -m '#{COMMIT_DESCRIPTION}' Gemfile.lock")
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def git_push_origin
+  ErrorEmittingExecutor.execute("git push origin -u #{BRANCH_NAME}")
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def create_pr
+  ErrorEmittingExecutor.execute("hub pull-request -f -m '#{COMMIT_DESCRIPTION}'")
+end
+
+# Check the result string of ErrorEmittingExecutor command for failure,
+#   output a message if so
+def result_failure?(step_name, result)
+  if result == 'FAILED'
+    p "stopped after #{step_name} due to failure"
+    return true
+  end
+  false
+end
+
+def clone_or_update_repo(repo)
+  puts "\n**** Processing #{repo} ****"
+  if File.exist?(cloned_repo_dir(repo))
+    update_repo(repo)
+  else
+    create_repo(repo)
+  end
+end
+
+def update_repo(repo)
+  within_cloned_repo_dir(repo) do
+    fetch_update
+    # we need local main branch to be up to date if we are creating a branch based on it
+    ErrorEmittingExecutor.execute('git switch main', exit_on_error: true)
+    # This allows our default branches to vary across projects
+    ErrorEmittingExecutor.execute('git reset --hard $(git symbolic-ref refs/remotes/origin/HEAD)',
+                                  exit_on_error: true)
+    ErrorEmittingExecutor.execute('bundle install', exit_on_error: true)
+  end
+end
+
+# fetch main and our latest branch from origin if it exists
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def fetch_update
+  remote_branch = ErrorEmittingExecutor.execute("git branch -r --list origin/#{BRANCH_NAME}")
+  if remote_branch == 'origin/cocina-level2-updates'
+    ErrorEmittingExecutor.execute("git fetch origin #{GIT_MAIN_FETCH_REFS} #{GIT_BRANCH_FETCH_REFS}")
+  else
+    ErrorEmittingExecutor.execute("git fetch origin #{GIT_MAIN_FETCH_REFS}", exit_on_error: true)
+  end
```

Comment:
Curious if it's possible to fetch all branches every time, but maybe this logic gets us something that I'm missing.

Reply:
we can fetch all the branches and simplify this logic, but ... there's no real need, and if we're going to the trouble of only doing a depth of 5, I figured the clone in the workspace could be tailored to this scripts needs.  Could be revisited if this script becomes more general use.   Also, "*" gets all the tags in addition to all active branches.  I'm not sure how much this impacts the size of the cloned directory.

Reply:
OK 🤷🏻‍♂️